lib/ansible/modules/monitoring/zabbix/zabbix_action.py

```diff
@@ -794,15 +800,15 @@ def _construct_parameters(self, **kwargs):
                 'discovery',
                 'auto_registration',
                 'internal'], kwargs['event_source']),
-            'esc_period': kwargs.get('esc_period'),
-            'filter': kwargs['conditions'],
-            'def_longdata': kwargs['default_message'],
-            'def_shortdata': kwargs['default_subject'],
-            'r_longdata': kwargs['recovery_default_message'],
-            'r_shortdata': kwargs['recovery_default_subject'],
-            'ack_longdata': kwargs['acknowledge_default_message'],
-            'ack_shortdata': kwargs['acknowledge_default_subject'],
-            'operations': kwargs['operations'],
+            'esc_period': kwargs['esc_period'],
+            'filter': kwargs.get('conditions'),
+            'def_longdata': kwargs.get('default_message'),
+            'def_shortdata': kwargs.get('default_subject'),
+            'r_longdata': kwargs.get('recovery_default_message'),
+            'r_shortdata': kwargs.get('recovery_default_subject'),
+            'ack_longdata': kwargs.get('acknowledge_default_message'),
+            'ack_shortdata': kwargs.get('acknowledge_default_subject'),
+            'operations': kwargs.get('operations'),
```

Comment:
Every module argument defaults to `None` if not overridden via `default=something` in `argument_spec` so this change may be unnecessary at all, but I am be missing something here that you see.

Reply:
Maybe you're right and this change is not unnecessary. The reason why I made it was because if I submit and empty field that is defined as `kwargs['something']`, an exception will be risen.
Seems like Zabbix API need any type of operations to be present in request: operations/recovery operations/acknowledge operations. I'm not that good at Python, so I've decided not to check if any of those fields are present in config and just let Zabbix API to handle errors.
Could you advise me how this problem can be solved? The only thing I wanted to fix is validation that makes 'operations' field mandatory although mandatory fields are conditional and only one of them must present.
So, I'm open for suggestions from more experienced developers. :)

Reply:
@vzabawski can you paste the exception you are getting please? Your change may be valid, but I am still having trouble understanding it fully.

@rubentsirunyan what's your opinion on this?

Reply:
This may get overriden with what I've mentioned in other [comment](https://github.com/ansible/ansible/pull/58379/files#r301533237) as this module wouldn't let you use itself without `esc_period` argument.

For the other arguments (changed on lines 804 - 811) change from `kwargs['XYZ']` to `kwargs.get('XYZ')` will still return `None` as default.

Reply:
Code:
```
- name: Test action
  zabbix_action:
    server_url: "{{ zabbix_url }}"
    login_user: "{{ zabbix_login }}"
    login_password: "{{ zabbix_password }}"
    name: "Test action"
    event_source: trigger
    esc_period: 3600
    state: present
    status: enabled

    eval_type: "and"

    conditions:
      - type: trigger_name
        operator: "like"
        value: "memory"
    
    acknowledge_operations:
      - type: send_message
        media_type: 'Email'
        send_to_users:
          - 'Admin'
```
Error:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'NoneType' object is not iterable
fatal: [127.0.0.1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2070, in <module>\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2061, in main\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 1104, in construct_the_data\nTypeError: 'NoneType' object is not iterable\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

Reply:
Now I install patched version:
```
valery@valery-VirtualBox:~$ ansible --version
ansible 2.9.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/valery/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15+ (default, Nov 27 2018, 23:36:35) [GCC 7.3.0]
```
Output (partial):
```
TASK [Test action] *************************************************************
changed: [127.0.0.1]

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```
The action appears in Zabbix.

Reply:
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'NoneType' object is not iterable
fatal: [127.0.0.1]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/valery/.ansible/tmp/ansible-tmp-1562674167.5-10943360351739/AnsiballZ_zabbix_action.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2070, in <module>\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 2061, in main\n  File \"/tmp/ansible_zabbix_action_payload_gTRlch/__main__.py\", line 1104, in construct_the_data\nTypeError: 'NoneType' object is not iterable\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

Actually, this error has nothing to do with this change (`kwargs[]` or `kwargs.get()`). So @D3DeFi is right, there is no need to use `kwargs.get()`, as not specified arguments default to `None`.
The reason why your patch does not throw this error is this change you made:
https://github.com/ansible/ansible/blob/62b3fec7055cfb1eef6b95cace8dd324a16fd97b/lib/ansible/modules/monitoring/zabbix/zabbix_action.py#L1116-L1117
Without it, if you try pass `None` to the method it tries to iterate over it which obviously fails with `TypeError: 'NoneType' object is not iterable`

I think defaulting all the arguments that have `type=list` to `[]` is a prettier solution to this issue.
What do you think?

Reply:
Agree, explicit things are usually more readable and cleaner solution. Btw, can you drop changelog fragment to [this folder as well](https://github.com/ansible/ansible/tree/devel/changelogs/fragments) please?

No need to document that we are changing defaults from `None` to `[]`, but information like: "`esc_period` is now required to reflect actual Zabbix API call" and "support for new condition operators (x, y, z) added in Zabbix 4.2" is a good info to have in release changelog, when this hits stable release.

Thank you both for all the work so far

Reply:
@D3DeFi So, could you please clarify what should be done from my side? I guess I need to revert changes made to arguments and leave the following fragment:
```
if operations is None: 
return None 
```

Am I right? Btw, should I squash all commits into one?

Reply:
1. `esc_period` in `argument_spec` should have `required=True` instead of `False` to match your documentation change
2. changes on lines 803 - 811 from your patch should be reverted (`kwargs[]` vs `kwargs.get()`)
3. argument options `conditions` (line 1690), `operations` (line 1704), `recovery_operations` (line 1812) and `acknowledge_operations` (line 1896) should have `default=[]` parameter as they are `type=list`
4. changelog fragment should be added to ansible-repo/changelog/fragments directory (see my earlier comment in this thread)

Reply:
And yes, it would be cool if you could squash all of the commits into one!